webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala

```diff
@@ -212,30 +219,32 @@ trait AnalyseInterpolation[T, I <: GenericInterpolationContext[T]] {
    * The funny part that for each token adds the interpretation of the token
    * by composing interpretation function.
    */
-  def parseToken(tokens:List[Token]): I => PureResult[String] = {
+  def parseToken(tokens:List[Token], propertyEngineService: PropertyEngineService): I => PureResult[String] = {
     def build(context: I) = {
       val init: PureResult[String] = Right("")
       (tokens).foldLeft(init) {
-        case (Right(str), token) => analyse(context, token).map(s => (str + s))
+        case (Right(str), token) => analyse(context, token, propertyEngineService).map(s => (str + s))
         case (Left(err) , _    ) => Left(err)
       }
     }
 
     build _
   }
 
-  /*
+  /*u4i
+
    * The three following methods analyse token one by one and
    * given the token, build the function to execute to get
    * the final string (that may not succeed at run time, because of
    * unknown parameter, etc)
    */
-  def analyse(context: I, token:Token): PureResult[String] = {
+  def analyse(context: I, token:Token, propertyEngineService: PropertyEngineService): PureResult[String] = {
     token match {
       case CharSeq(s)          => Right(s)
       case NonRudderVar(s)     => Right(s"$${${s}}")
       case NodeAccessor(path)  => getNodeAccessorTarget(context, path)
       case Param(path)         => getRudderGlobalParam(context, path)
+      case RudderEngine(e, m, p) => Right(ZioRuntime.runNow(propertyEngineService.process(e, m, p)))
```

Comment:
You don't have the right to do that, you are lying to code caller: the code it not pure, it actually runs a (side!) effect, and says it's always OK - here, an error will lead to an exception. 
You should have a `expandWithPropertyEngine(context: I, engineName: String (= e), nameSpace: List[String] (= m), params: ConfigValue (= p)): IOResult[String]`.
It will taint the whole thing to `IOResult`, which means that we go from "we have something that is pure/whitout side effects" to "only gods know what happen. Can block, call network, etc". 
We don't like `IOResult`, they are too powerfull. But most of the time, we don't have a choice.

Reply:
Not sure we want to push everything to ZIO World here, we could but i think we can keep to PureResult here....

So either we are going like François said and we go to IOResult all the way down (until we need to ZIORuntime.runNow) but I think we may keep to pure Result here by running it into an either

like this ZioRuntime.runNow(io.either) 


But in all cases, like François said you can't put it in a Right like that here,

Reply:
Maybe the signature should be 

def expandRudderEngineProperty (context : I, rudderEngine : RudderEngine) : IOResult/PureResult[String] 

to express that this function should only be used with a RudderEngine property, and not something else